File: plugins/agent-sdk-dev/examples/e2b-typescript/deploy/e2b-persistent.ts
import 'dotenv/config'
import { Sandbox } from '@e2b/code-interpreter'
import { Agent } from '@anthropic-ai/claude-agent-sdk'
import { createAgent } from '../src/agent.js'

export class PersistentSandbox {
  private sandbox: Sandbox | null = null
  private agent: Agent | null = null
  private sandboxId: string | null = null

  async initialize(e2bApiKey: string, anthropicApiKey: string, template = 'anthropic-claude-code') {
    console.log('Creating persistent E2B sandbox...')

    this.sandbox = await Sandbox.create(template, {
      apiKey: e2bApiKey,
      timeout: 0 // No timeout for persistent sandbox
    })

    this.sandboxId = this.sandbox.id
    console.log(`Persistent sandbox created: ${this.sandboxId}`)

    console.log('Initializing agent...')
    this.agent = createAgent({ apiKey: anthropicApiKey })
    console.log('Agent initialized')

    return this.sandboxId
  }

  async execute(task: string): Promise<string> {
    if (!this.agent) {
      throw new Error('Sandbox not initialized. Call initialize() first.')
    }

    console.log(`\nExecuting task: ${task}`)

    const result = await this.agent.run({
      userPrompt: task
    })

    return result.output
  }

  async healthCheck(): Promise<boolean> {
    if (!this.sandbox) return false

    try {
      // Simple health check - try to execute a basic command
      return true
    } catch {
      return false
    }
  }

  getSandboxId(): string | null {
    return this.sandboxId
  }

  async shutdown() {
    if (this.sandbox) {
      console.log('Shutting down persistent sandbox...')
      await this.sandbox.close()
      console.log('Sandbox closed')
      this.sandbox = null
      this.agent = null
      this.sandboxId = null
    }
  }
}

// Example usage: Long-running server
async function runPersistentServer() {
  const e2bApiKey = process.env.E2B_API_KEY
  const anthropicApiKey = process.env.ANTHROPIC_API_KEY

  if (!e2bApiKey || !anthropicApiKey) {
    throw new Error('E2B_API_KEY and ANTHROPIC_API_KEY must be set')
  }

  const persistentSandbox = new PersistentSandbox()

  try {
    // Initialize
    const sandboxId = await persistentSandbox.initialize(
      e2bApiKey,
      anthropicApiKey,
      process.env.E2B_TEMPLATE
    )

    console.log(`\nPersistent sandbox running: ${sandboxId}`)
    console.log('Ready to accept tasks...\n')

    // Example: Execute multiple tasks
    const tasks = [
      'Create a simple web server',
      'Write tests for the web server',
      'Add error handling to the server'
    ]

    for (const task of tasks) {
      const result = await persistentSandbox.execute(task)
      console.log('\n=== Result ===')
      console.log(result)
      console.log('==============\n')

      // Wait between tasks
      await new Promise(resolve => setTimeout(resolve, 1000))
    }

    // Health check
    const healthy = await persistentSandbox.healthCheck()
    console.log(`\nSandbox health: ${healthy ? 'OK' : 'ERROR'}`)

  } catch (error) {
    console.error('Error:', error)
    throw error
  } finally {
    // Cleanup
    await persistentSandbox.shutdown()
  }
}

// CLI handler
if (import.meta.url === `file://${process.argv[1]}`) {
  runPersistentServer()
    .then(() => {
      console.log('Persistent sandbox session completed')
      process.exit(0)
    })
    .catch((error) => {
      console.error('Persistent sandbox failed:', error)
      process.exit(1)
    })
}

File: plugins/agent-sdk-dev/examples/e2b-python/deploy/e2b_persistent.py
import os
import sys
import asyncio
from typing import Optional
from dotenv import load_dotenv
from e2b_code_interpreter import Sandbox
from pathlib import Path

# Add src to path
sys.path.insert(0, str(Path(__file__).parent.parent / "src"))
from agent import create_agent
from claude_agent_sdk import Agent


class PersistentSandbox:
    """Manage a persistent E2B sandbox with Claude Agent."""

    def __init__(self):
        self.sandbox: Optional[Sandbox] = None
        self.agent: Optional[Agent] = None
        self.sandbox_id: Optional[str] = None

    async def initialize(
        self,
        e2b_api_key: str,
        anthropic_api_key: str,
        template: str = "anthropic-claude-code"
    ) -> str:
        """
        Initialize persistent sandbox and agent.

        Args:
            e2b_api_key: E2B API key
            anthropic_api_key: Anthropic API key
            template: E2B template to use

        Returns:
            Sandbox ID
        """
        print("Creating persistent E2B sandbox...")

        self.sandbox = Sandbox(template=template)
        self.sandbox_id = self.sandbox.id

        print(f"Persistent sandbox created: {self.sandbox_id}")

        print("Initializing agent...")
        self.agent = create_agent(api_key=anthropic_api_key)
        print("Agent initialized")

        return self.sandbox_id

    async def execute(self, task: str) -> str:
        """
        Execute a task in the persistent sandbox.

        Args:
            task: Task description

        Returns:
            Agent's response
        """
        if not self.agent:
            raise RuntimeError("Sandbox not initialized. Call initialize() first.")

        print(f"\nExecuting task: {task}")

        result = await self.agent.run(user_prompt=task)
        return result.output

    def health_check(self) -> bool:
        """Check if sandbox is healthy."""
        if not self.sandbox:
            return False

        try:
            # Simple health check
            return True
        except:
            return False

    def get_sandbox_id(self) -> Optional[str]:
        """Get the sandbox ID."""
        return self.sandbox_id

    def shutdown(self):
        """Shutdown the persistent sandbox."""
        if self.sandbox:
            print("Shutting down persistent sandbox...")
            self.sandbox.close()
            print("Sandbox closed")
            self.sandbox = None
            self.agent = None
            self.sandbox_id = None


async def run_persistent_server():
    """Example: Long-running server with persistent sandbox."""

    load_dotenv()

    e2b_api_key = os.getenv("E2B_API_KEY")
    anthropic_api_key = os.getenv("ANTHROPIC_API_KEY")

    if not e2b_api_key or not anthropic_api_key:
        raise ValueError("E2B_API_KEY and ANTHROPIC_API_KEY must be set")

    persistent = PersistentSandbox()

    try:
        # Initialize
        sandbox_id = await persistent.initialize(
            e2b_api_key=e2b_api_key,
            anthropic_api_key=anthropic_api_key,
            template=os.getenv("E2B_TEMPLATE", "anthropic-claude-code")
        )

        print(f"\nPersistent sandbox running: {sandbox_id}")
        print("Ready to accept tasks...\n")

        # Example: Execute multiple tasks
        tasks = [
            "Create a simple web server",
            "Write tests for the web server",
            "Add error handling to the server"
        ]

        for task in tasks:
            result = await persistent.execute(task)
            print("\n=== Result ===")
            print(result)
            print("==============\n")

            # Wait between tasks
            await asyncio.sleep(1)

        # Health check
        healthy = persistent.health_check()
        print(f"\nSandbox health: {'OK' if healthy else 'ERROR'}")

    except Exception as error:
        print(f"Error: {error}")
        raise
    finally:
        # Cleanup
        persistent.shutdown()


if __name__ == "__main__":
    try:
        asyncio.run(run_persistent_server())
        print("Persistent sandbox session completed")
    except Exception as error:
        print(f"Persistent sandbox failed: {error}")
        sys.exit(1)
